### PR TITLE
refactor YOML attribute parsing

### PR DESCRIPTION
**Pull request by kazuho**
Until now, every configuration directive that (optionally) takes a YOML mapping has written it's own attribute parser.

The fact has not only made the code-base unnecessary complex but also has caused inconsistency between the parsers in how they deal with unknown or duplicate key.

This PR aims to resolve the issue by introducing a function to be used for fetching named values from a YOML mapping.

rework of #1068

### CODE REVIEW

**Comment by kazuho** [comment]
@i110 Could you please review the PR?

File: lib/handler/configurator/proxy.c [hunk 1]
```diff
@@ -268,29 +252,23 @@ static int on_config_ssl_session_cache(h2o_configurator_command_t *cmd, h2o_conf
...
     case YOML_TYPE_MAPPING: {
-        yoml_t *weight_node;
-        if ((url_node = yoml_get(backend, "url")) == NULL) {
-            h2o_configurator_errprintf(cmd, backend, "cannot find mandatory property `url`");
-            return NULL;
-        }
-        if (url_node->type != YOML_TYPE_SCALAR) {
-            h2o_configurator_errprintf(cmd, url_node, "value of the `url` property must be a scalar");
+        yoml_t **weight_node;
+        if (h2o_configurator_parse_mapping(cmd, backend, "url:s", "weight", &url_node, &weight_node) != 0)
```
Comment (i110) [inline lib/handler/configurator/proxy.c]: `weight` must be scalar?

File: lib/handler/configurator/errordoc.c [hunk 2]
```diff
@@ -46,84 +46,57 @@ static int scan_and_check_status(h2o_configurator_command_t *cmd, yoml_t *value,
...
-                    for (k = 0; k != status_len; ++k) {
-                        if (status[k] == parsed) {
-                            h2o_configurator_errprintf(cmd, value, "status %d appears multiple times", status[k]);
-                            return -1;
-                        }
-                    }
-                    status[status_len++] = parsed;
-                }
-            } else {
-                if (scan_and_check_status(cmd, value, &parsed) != 0)
```
[the remark this answers is not in the document]
Comment (kazuho) [reply lib/handler/configurator/errordoc.c]: Addressed in d883788.

**Comment by kazuho** [comment]
@i110 Thank you for the review.

> Actually both of `capacity` and `lifetime` must be scalars?
> `weight` must be scalar?

The types are actually checked later when `h2o_configurator_scanf` is applied. The intention here is to emit more specific error that states the required type rather than just saying that it should be a scalar.

But I agree that such coding convention is confusing. Therefore, you suggestion

> IMHO it's better to explicitly declare whether each directives expect a scalar or not.

sounds good to me.

How about adding prefix `:a` (meaning _any_) to explicitly indicate that the check is not done in `h2o_configurator_parse_mappings`, and use that for the attributes that are later checked using `h2o_configurator_scanf`?

**Comment by kazuho** [comment]
With the commits above, `h2o_configurator_parse_mappings` now enforces the use of the type specifiers. One or more specifiers shown below must be associated to each attribute.

* `s` - scalar
* `a` - sequence (a.k.a. array)
* `m` - mapping
* `*` - any of the above

**Pull request merged**